hub/values.yaml

```diff
@@ -141,8 +148,68 @@ jupyterhub:
               line = f.readline()
           return members
 
+        def course_error_messages(courses):
+          return map(lambda x: x.get('message', ''), courses['errors'])
+
+        def course_filter(courses, enrollment_type, attr='id'):
+          '''Filter course attributes by enrollment type.'''
+          attrs = []
+          for course in courses:
+            if 'enrollments' not in course: continue
+            for enrollment in course['enrollments']:
+              if enrollment['type'] == enrollment_type:
+                attrs.append(course[attr]); break
+          return attrs
+
+        class InvalidAccessToken(Exception): pass
+
         class CustomAttrSpawner(KubeSpawner):
 
+          canvas_client_id = Unicode('', config=True,
+            help="OAuth client id to refresh the access token."
+          )
+          canvas_client_secret = Unicode('', config=True,
+            help="OAuth client secret to refresh the access token."
+          )
+          canvas_url = Unicode('', config=True,
+            help="Canvas URL."
+          )
+
+          async def refresh_access_token(self, auth_state):
+            '''Use the user's refresh token to get a new access token.'''
+            url = f"{self.canvas_url}/login/oauth2/token"
+            params = dict(
+              grant_type = 'refresh_token',
+              client_id = self.canvas_client_id,
+              client_secret = self.canvas_client_secret,
+              refresh_token = auth_state['refresh_token']
+            )
+            data = {}
+            async with aiohttp.ClientSession() as session:
+              async with session.post(url, params=params) as r:
+                data = await r.json()
+            auth_state['access_token'] = data['access_token']
+            await self.user.save_auth_state(auth_state)
+            return data['access_token']
+
+          @lru_cache(maxsize=1024)
+          async def get_canvas_courses(self, token, enrollment_type=None):
```

Comment:
CanvasAuthenticator is also something we write customly. Can we move this to CanvasAuthenticator? We can then put the following code in CanvasAuthenticator:

1. Fetching courses
2. Dealing with expired tokens
3. Throwing errors about expired tokens

That lets us keep *just* the stuff about policy enforcment - profiles, admin determination, etc, here.

We should also move this spawner out of the YAML into the hub image, but that can wait.

Reply:
You can then access the authenticator from the spawner as 'self.authenticator' (See https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/spawner.py#L150)

Reply:
This will also make it easier for us to upstream CanvasAuthenticator more easily.

Reply:
This will also make it easier for us to upstream CanvasAuthenticator more easily.

Reply:
Oh, I see, the course fetching will be invoked by the spawner, but the code will live in the authenticator?

Reply:
Yeah! I think of this code as providing 'groups' the user belongs to, and would like to push that upstream eventually. The authenticator provides this information. Various other parts can do whatever they want with this info. In our case, our spawner is doing some complex stuff to figure out who gets how much RAM...